### PR TITLE
Replace remaining dynamic_cast / typeid

### PR DESCRIPTION
**Pull request by brunoabinader**
Adding `-fno-rtti` is currently blocked because of some Boost headers:
```
In file included from ../../../mason_packages/headers/boost/1.65.1/include/boost/iostreams/stream_buffer.hpp:21:0,
                 from ../../../mason_packages/headers/boost/1.65.1/include/boost/iostreams/stream.hpp:21,
                 from ../../../mason_packages/headers/boost/1.65.1/include/boost/spirit/home/qi/stream/detail/iterator_source.hpp:14,
                 from ../../../mason_packages/headers/boost/1.65.1/include/boost/spirit/home/qi/stream/stream.hpp:16,
                 from ../../../mason_packages/headers/boost/1.65.1/include/boost/spirit/home/qi/stream.hpp:15,
                 from ../../../mason_packages/headers/boost/1.65.1/include/boost/spirit/home/qi.hpp:30,
                 from ../../../mason_packages/headers/boost/1.65.1/include/boost/spirit/include/qi.hpp:16,
                 from ../../../src/mbgl/text/language_tag.cpp:10:
../../../mason_packages/headers/boost/1.65.1/include/boost/iostreams/detail/streambuf/direct_streambuf.hpp: In member function ‘const std::type_info& boost::iostreams::detail::direct_streambuf<T, Tr>::component_type() const’:
../../../mason_packages/headers/boost/1.65.1/include/boost/iostreams/detail/streambuf/direct_streambuf.hpp:77:67: error: cannot use typeid with -fno-rtti
     const std::type_info& component_type() const { return typeid(T); }
                                                                   ^
In file included from ../../../mason_packages/headers/boost/1.65.1/include/boost/iostreams/stream_buffer.hpp:22:0,
                 from ../../../mason_packages/headers/boost/1.65.1/include/boost/iostreams/stream.hpp:21,
                 from ../../../mason_packages/headers/boost/1.65.1/include/boost/spirit/home/qi/stream/detail/iterator_source.hpp:14,
                 from ../../../mason_packages/headers/boost/1.65.1/include/boost/spirit/home/qi/stream/stream.hpp:16,
                 from ../../../mason_packages/headers/boost/1.65.1/include/boost/spirit/home/qi/stream.hpp:15,
                 from ../../../mason_packages/headers/boost/1.65.1/include/boost/spirit/home/qi.hpp:30,
                 from ../../../mason_packages/headers/boost/1.65.1/include/boost/spirit/include/qi.hpp:16,
                 from ../../../src/mbgl/text/language_tag.cpp:10:
../../../mason_packages/headers/boost/1.65.1/include/boost/iostreams/detail/streambuf/indirect_streambuf.hpp: In member function ‘const std::type_info& boost::iostreams::detail::indirect_streambuf<T, Tr, Alloc, Mode>::component_type() const’:
../../../mason_packages/headers/boost/1.65.1/include/boost/iostreams/detail/streambuf/indirect_streambuf.hpp:97:67: error: cannot use typeid with -fno-rtti
     const std::type_info& component_type() const { return typeid(T); }
```

/cc @mapbox/gl-core @mapbox/embedded

### CODE REVIEW

**Comment by jfirebaugh** [comment]
I tried replacing the uses of boost::spirit with no-ops just to be able to compile and get a sense of the effect on binary size `-fno-rtti` would have. Since it removes some code, this will overstate the gains a bit, but it seems promising nonetheless:

```
     VM SIZE                                FILE SIZE
 --------------                          --------------
 +23e2% +2.45Ki [__LINKEDIT]                  -8 -66.7%
   +22%    +508 [__TEXT]                    +508   +22%
  -0.8%     -32 Indirect Symbol Table        -32  -0.8%
  -0.0%     -41 __TEXT,__cstring             -41  -0.0%
  -0.5%     -72 Lazy Binding Info            -72  -0.5%
  -0.8%     -88 [4 Others]                   -88  -0.8%
  -1.0%    -160 Symbol Table                -160  -1.0%
  -1.1%    -200 Function Start Addresses    -200  -1.1%
  -2.1%    -208 Export Info                 -208  -2.1%
  -0.4%    -308 __TEXT,__unwind_info        -308  -0.4%
 -20.5%    -424 Weak Binding Info           -424 -20.5%
  -0.1%    -488 __TEXT,__gcc_except_tab     -488  -0.1%
  -3.1%    -864 String Table                -864  -3.1%
 -42.9%    -928 [__DATA]                    -928 -24.5%
 -27.6% -1.35Ki Rebase Info              -1.35Ki -27.6%
  -2.2% -1.38Ki Code Signature           -1.38Ki  -2.2%
 -18.6% -1.80Ki Binding Info             -1.80Ki -18.6%
 -28.5% -8.12Ki __DATA,__data            -8.12Ki -28.5%
  -0.3% -9.80Ki __TEXT,__text            -9.80Ki  -0.3%
 -28.6% -10.9Ki __DATA,__const           -10.9Ki -28.6%
 -31.3% -73.8Ki __TEXT,__const           -73.8Ki -31.3%
  -2.5%  -108Ki TOTAL                     -110Ki  -2.6%
```

File: include/mbgl/util/unique_any.hpp [hunk 1]
```diff
@@ -1,17 +1,19 @@
 #pragma once
```
Comment (jfirebaugh) [inline include/mbgl/util/unique_any.hpp]: Let's just replace `unique_any` with [this code](https://github.com/mapbox/mapbox-bindgen/blob/master/src/support/mapbox/bindgen/peer.hpp) from mapbox-bindgen. (For now we can just copy it.)

File: src/mbgl/tile/tile.hpp [hunk 2]
```diff
@@ -35,7 +35,13 @@ class Context;
 
 class Tile : private util::noncopyable {
 public:
-    Tile(OverscaledTileID);
+    enum class Kind : int32_t {
```
Comment (asheemmamoowala) [inline src/mbgl/tile/tile.hpp]: nit: `uint8_t` should suffice right?

**Comment by brunoabinader** [comment]
@ChrisLoer It seems `-DBOOST_SPIRIT_X3_NO_RTTI` is not enough for some of our Boost header dependencies e.g. I still catch [these same errors](https://github.com/mapbox/mapbox-gl-native/pull/12475#issue-203628363). Do you see an alternative for Boost Spirit in `src/mbgl/text/language_tag.cpp`? Can we use ICU helper functions for that?

**Comment by ChrisLoer** [comment]
@brunoabinader Post-API level 21, Android supports BCP 47 natively, so it's really only for pre-21 that we need the spirit-based language tag (on non-Apple/Android platforms, we don't support locales in `Collator` so it's not necessary). I'm not sure if it's really practical to make two separate Android builds to support different API levels, though?

ICU does have its own BCP 47 parsing code `uloc_forLanguageTag`, but we'd also have to figure out how to go from an ICU locale ID back to the ISO language/region/script codes (`uloc_getISO3Language` looks promising, although I don't immediately see a way to get ISO script tags back out). I think the bigger barrier to using ICU is that it would require linking some substantial chunk of the ICU locale code, which we're not currently using. It's worth an experiment, but I suspect it might undo much of any binary size benefit we're looking for.

Another option might just be to do a quick-and-dirty parsing of a subset of the BCP 47 language, similar to what Firefox does: https://dxr.mozilla.org/mozilla-central/source/intl/locale/MozLocale.cpp. It would feel like a step backward to go from "we understand BCP 47" to "it's like BCP 47 and you probably won't have any issues but if you ever do it's going to take us a while to figure out what's going on". That said, if it's good enough for Firefox that might be a sign that it should be good enough for us.

**Comment by jfirebaugh** [comment]
@brunoabinader We are hoping that removing RTTI decreases binary size, but this actually increased it, I suspect due to pulling in `std::regex`.

image

Since we still have a Boost.Spirit dependency anyway, can we revert that part of this PR for now and maybe try a hand-written parser?

**Comment by brunoabinader** [comment]
> Since we still have a Boost.Spirit dependency anyway, can we revert the CacheControl::parse() part of this PR for now and maybe try a hand-written parser?

Sure :+1: will do that on a separate PR - good catch!

**Pull request merged**